### PR TITLE
#19. Moves the regex for determining the id into the config.

### PR DESCRIPTION
**Pull request by dncnmcdougall**
In particular all three regex expressions used in browse have been moved to config.
I also added options to specify whether to use the filename and the title to determine the id.
The default is to use both. When both are used they are checked against each other. If they differ an error is raised.

I store my ids only in the filename, which is why I introduced this.

### CODE REVIEW

**Comment by dncnmcdougall** [comment]
Because I don't put my id's in the title and I use a very general regex for the id ("%w+") I got very weird results for the id initially. This is another reason why I implemented the explicit flags.

**Comment by dncnmcdougall** [comment]
Also I have been building my own zettlekasten code (as one does) and thought I would look at your vim plugin to use ideas from it for my front end. But the more I looked the more I really liked your code and so have decided to use it directly. This is really well written. And full! Thank you.

File: lua/zettelkasten/browser.lua
Comment (Furkanzmc) [inline lua/zettelkasten/browser.lua:11]: [non-blocking] You could just pass the `filename_pattern` here instead of passing the entire `config` object. It exposes more than this function needs and makes it just slightly harder to maintain since it's possible that this function could internally change the `config` object. Or you could just fetch the `config` internally here without relying on a parameter.

For places where multiple items from the config is accessed, and you can pass a new table with just the appropriate ones. If there isn't a lot, then just adding 1-2 more arguments is fine.

File: lua/zettelkasten/browser.lua
Comment (Furkanzmc) [inline lua/zettelkasten/browser.lua:20]: [non-blocking] Same as above.

File: lua/zettelkasten/browser.lua
Comment (Furkanzmc) [inline lua/zettelkasten/browser.lua:44]: This prevents the browser from being loaded. I'm slightly worried about the effect this might have when there's multiple files with mismatching `id` in the title and the file name. Maybe you can append something like `[ID-MISMATCH]` to the title of the note so that the browser loads all the notes and one can easily see all the notes with the mismatch and correct them. Otherwise when there's more than one, it'd be frustrating to fix.

File: lua/zettelkasten/browser.lua
Comment (Furkanzmc) [inline lua/zettelkasten/browser.lua:58]: [non-blocking] Same as the first comment. You can just pass `id_pattern` here.

**Review by Furkanzmc — CHANGES_REQUESTED** [review]
> This is another reason why I implemented the explicit flags.

I think those explicit flags are a good idea. However, I'm a bit concerned about allowing both `id_in_filename` and `id_in_title` to be set to `true`. I think we should allow just one. The reason is that what happens when one sets them both to `true` and wants to mix the usage of it where they have the `id` for a note in the file name and not the title and in others they have it in both or just in the title?

Setting them both to true gives the impression that they can be mixed but that's not what the implementation is doing. And I think allowing for both increases the complexity of the code for little gain.

> Also I have been building my own zettlekasten code (as one does) and thought I would look at your vim plugin to use ideas from it for my front end. But the more I looked the more I really liked your code and so have decided to use it directly. This is really well written. And full! Thank you.

I'm very happy to hear that! Thank you! I'm glad it was useful and saved you time. I appreciate the contribution, thanks for taking the time to make the plugin better!

-----

Thanks for the PR! Overall, it looks good. I don't have strong enough reasons to block the PR with the comments marked with `[non-blocking]` but do let me know your thoughts about it.

Can you also format your code using `stylua`, please?

**Comment by dncnmcdougall** [comment]
Excellent thank you for these comments. May I try and summarise:
1. don't allow `id_in_title` and `id_in_filename` to be true at the same time.
2. run `stylua`
3. don't pass the whole config dictionary in, rather either explicit parameters or silently (invisibly) get the config internally.
4. on mismatch of id's report the error more gracefully so that the plugin may still load.
Does this capture all your points? (It is wonderful having a good review, thank you for the effort.)

As a solution to 1. and 4. we could use an enum. Something like:
> {infer_id} (optional| integer) This is a flag to determine where the id should be inferred from. 
> Set to 0 it is inferred from the > title. Set to 1 it is inferred from the filename. 
> Default: 0

This keeps it simple in that there can no longer be a mismatch (because we never check both.) Of course that means that we do not check that these are in sync. But that is not done at the moment either.

3 is easy to fix. I think in this case explicit parameters will work best.
I will figure out 2. A styler is a boon worth investing in indeed.

**Comment by Furkanzmc** [comment]
> Excellent thank you for these comments. May I try and summarise:
> 
>     1. don't allow `id_in_title` and `id_in_filename` to be true at the same time.
> 
>     2. run `stylua`
> 
>     3. don't pass the whole config dictionary in, rather either explicit parameters or silently (invisibly) get the config internally.
> 
>     4. on mismatch of id's report the error more gracefully so that the plugin may still load.
>        Does this capture all your points? (It is wonderful having a good review, thank you for the effort.)
> 

This summarizes it perfectly!

> As a solution to 1. and 4. we could use an enum. Something like:
> 
> > {infer_id} (optional| integer) This is a flag to determine where the id should be inferred from.
> > Set to 0 it is inferred from the > title. Set to 1 it is inferred from the filename.
> > Default: 0
> 
> This keeps it simple in that there can no longer be a mismatch (because we never check both.) Of course that means that we do not check that these are in sync. But that is not done at the moment either.

I think that's a good solution. Since we use the file names for jumping between different files, the mismatch for the `id` in the title and the `id` in the file name shouldn't matter.

> 3 is easy to fix. I think in this case explicit parameters will work best.

Thank you.

> I will figure out 2. A styler is a boon worth investing in indeed.

All you need to do is install [stylua](https://github.com/JohnnyMorganz/StyLua) and just run `:!stylua %` in Vim. It'll pick up the appropriate configuration file automatically.

File: lua/zettelkasten/browser.lua
Comment (Furkanzmc) [inline lua/zettelkasten/browser.lua:33]: Can you assign this `0` and `1` values to some named constants?

**Review by Furkanzmc — APPROVED** [review]
Looks great! Thank you for the contribution. I only have one comment. Once you make that change, it'll be good to merge. Thank you!

File: lua/zettelkasten/browser.lua
Comment (dncnmcdougall) [reply lua/zettelkasten/browser.lua:33]: I don't know if lua can do actual enums. Would using string values suffice?

Comment (dncnmcdougall) [reply lua/zettelkasten/browser.lua]: Actually I created confid.TITLE and config.FILENAME and used those. instead of a string.

**Review by Furkanzmc — APPROVED** [review]
Looks great. Thank you!

**Comment by dncnmcdougall** [comment]
Excellent it has been a pleasure working on this with you.

**Pull request merged**
